pkg/cspc/algorithm/select_node.go

```diff
@@ -0,0 +1,222 @@
+limitations under the License.
+*/
+
+package algorithm
+
+import (
+	"github.com/openebs/maya/pkg/volume"
+	"github.com/pkg/errors"
+	openebsio "github.com/openebs/api/pkg/apis/openebs.io/v1alpha1"
+	"github.com/openebs/api/pkg/apis/types"
+	cstor "github.com/openebs/api/pkg/apis/cstor/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+// SelectNode returns a node where pool should be created.
+func (ac *Config) SelectNode() (*cstor.PoolSpec, string, error) {
+	usedNodes, err := ac.GetUsedNode()
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "could not get used nodes list for pool creation")
+	}
+	for _, pool := range ac.CSPC.Spec.Pools {
+		// pin it
+		pool := pool
+		nodeName, err := ac.GetNodeFromLabelSelector(pool.NodeSelector)
+		if err != nil || nodeName == "" {
+			klog.Errorf("could not use node for selectors {%v}", pool.NodeSelector)
+			continue
+		}
+		if ac.VisitedNodes[nodeName] {
+			continue
+		} else {
+			ac.VisitedNodes[nodeName] = true
+
+			if !usedNodes[nodeName] {
+				return &pool, nodeName, nil
+			}
+		}
+
+	}
+	return nil, "", errors.New("no node qualified for pool creation")
+}
+
+// GetNodeFromLabelSelector returns the node name selected by provided labels
+func (ac *Config)GetNodeFromLabelSelector(labels map[string]string) (string, error) {
+	nodeList, err := ac.kubeclientset.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: getLabelSelectorString(labels)})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get node list from the node selector")
+	}
+	if len(nodeList.Items) != 1 {
+		return "", errors.Errorf("invalid no.of nodes %d from the given node selectors", len(nodeList.Items))
+	}
+	return nodeList.Items[0].GetLabels()[string(types.HostNameLabelKey)], nil
+}
+
+// getLabelSelectorString returns a string of label selector form label map to be used in
+// list options.
+func getLabelSelectorString(selector map[string]string) string {
+	var selectorString string
+	for key, value := range selector {
+		selectorString = selectorString + key + "=" + value + ","
+	}
+	selectorString = selectorString[:len(selectorString)-len(",")]
+	return selectorString
+}
+
+// GetUsedNode returns a map of node for which pool has already been created.
+func (ac *Config) GetUsedNode() (map[string]bool, error) {
+	usedNode := make(map[string]bool)
+	cspList, err := ac.
+		clientset.
+		CstorV1().
+		CStorPoolInstances(ac.Namespace).
+		List(
+		metav1.
+		ListOptions{LabelSelector: string(types.CStorPoolClusterLabelKey) + "=" + ac.CSPC.Name},
+		)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list already created csp(s)")
+	}
+	for _, cspObj := range cspList.Items {
+		usedNode[cspObj.Labels[string(types.HostNameLabelKey)]] = true
+	}
+	return usedNode, nil
+}
+
+// GetBDListForNode returns a list of BD from the pool spec.
+func GetBDListForNode(pool cstor.PoolSpec) []string {
```

Comment:
How about moving to API repo because it looks like tightly coupled with `PoolSpec`. IMO looks like a getter method from the spec.